### PR TITLE
Add new module to create/update a docker swarm.

### PR DESCRIPTION
**Pull request by tbouvet**
##### SUMMARY
This PR adds a new module, docker_swarm, which can be used to manage docker swarm (create a swarm, join, ...).
It supports **docker swarm mode**.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
docker_swarm

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
This module uses docker API.

_Example to create a new swarm:_
```
- hosts: manager
  tasks:
  - name: Create a new Swarm Cluster
    docker_swarm:
       state: "init"
       advertise_addr: "192.168.1.2"
```
_And the result is:_
```
TASK [debug] ***********************************************************************************************************
ok: [yvao0e70] => {
    "tokens": {
        "actions": [
            "This cluster is already a swarm cluster: kkvxvgjjzwfj5qnko9hqx3ut1"
        ],
        "changed": false,
        "failed": false,
        "result": "",
        "swarm_facts": {
            "JoinTokens": {
                "Manager": "SWMTKN-1-0fqcvn9ac71tkai1gkcgmpigxm2cievrr5blsugtly4ci1y7ot-22jh4kepfifiww46km6gj99dn",
                "Worker": "SWMTKN-1-0fqcvn9ac71tkai1gkcgmpigxm2cievrr5blsugtly4ci1y7ot-4ia579jj2l150sqbwvabinu0z"
            }
        }
    }
}
```

With this module, you can **create** a swarm, **join** a node to a swarm, **remove** or **leave** a node from a swarm.

### CODE REVIEW

**Comment by samdoran** [comment]
cc @chouseknecht

File: test/sanity/validate-modules/ignore.txt
Comment (mattclay) [inline test/sanity/validate-modules/ignore.txt]: The documentation and/or module should be fixed instead of adding additional ignores. If the issue is in a common docs fragment, that should be fixed in a separate PR first.

Comment (tbouvet) [reply test/sanity/validate-modules/ignore.txt]: @mattclay PR is done : https://github.com/ansible/ansible/pull/38378
When this PR will be merged, I will rebase this one.

Comment (tbouvet) [reply test/sanity/validate-modules/ignore.txt]: @mattclay Rebase done included PR https://github.com/ansible/ansible/pull/38378

**Comment by mattclay** [comment]
@tbouvet Can you include an integration test for this module?

**Comment by tbouvet** [comment]
@mattclay Integration tests to create/remove a swarm is done. I can not test all functions because I need several machines (> 1).

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: Ok, so long descriptions are not a good idea. Best practice is to do:
- First line is a short description of what the parameter is
- Subsequent lines for additional information on behaviour, or what different possible values mean

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: All descriptions, except the short_description end with a dot.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: No need to explicitly state that a parameter is not required. That's the default.
It helps find required parameters when skimming docs if only the required parameter have a required-key.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: No need to quote strings in YAML except in very specific cases.
We tend to only quote strings when it is required for consistency and educational purposes.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: This is not declarative as we normally expect from modules (unless they can't be made idempotent for some reason). Rather than specifying an action to undertake (imperative) we prefer the user states how it should be. Present, or absent. If this module manages swarms, we expect **present** means "create a swarm if it not already exists" or "modify swarm to new properties", and we expect **absent** means "remove swarm if it exists".

In your case it seems the module does not just manage the swarm, but also nodes in the swarm ? If you intend users to add/remove nodes to a swarm, maybe a second module **docker_swarm_node** is in order ?

**Comment by pilou-** [comment]
Hi! @ThomasSteinbach, (as maintainer of `cloud/docker`namespace) could you take a look at this pull request? 

@tbouvet: if namespace maintainers don't review this pull-request, a way to proceed is to put this onto IRC [core meeting agenda](https://github.com/ansible/community/issues/308) and show up at the [meeting to discuss](https://github.com/ansible/community/blob/master/meetings/README.md).

**Comment by tbouvet** [comment]
@agronholm @cove @dusdanig @joshuaconner @kassiansun @keitwb @olsaki @softzilla @zfil As a maintainer of a module in the namespace, have you some remarks please or is it possible to merge it ?

**Review by dagwieers — CHANGES_REQUESTED** [review]
Didn't finish the review yet.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: I think the problem will be the same if I create a second module just for the nodes because nodes (and managers) could have different states : present, leave (temporary absent) and remove (definitively absent). So if I resume I will have 3 states for the second module : add/leave/remove.
What is your suggestion ?

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: So there are two concerns:

1. the naming of states (add/remove/leave) which are imperative statements
2. the module is managing 2 different object classes

I know there are other modules that break on both accounts too. Not being a docker swarm expert, I don't know if I can come up with the best solution.

1. Maybe use states **present**/**absent**/**unregistered** ? Or is it an option to have **present**/**absent**, and a separate parameter **enabled**
2. The main question is whether there are additional properties to manage that affect the swarm, or affect the node. That would be a good reason to have it split.

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: BTW the 3 states confused me before, but from the examples it seems **init**/**leave** affects the swarm (no node is provided as argument) and **join**/**remove** affects the node as part of the swarm. But I find those states also confusing, maybe there is an issue in the examples ?

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/65881/47/tests

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: __init__ is just to create (initialize) a new swarm environment.
__join/leave/remove__ affects nodes.
I have write one example by state in the documentation.

Comment (maxamillion) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: I do agree that the states are actions instead of declared intent which is pretty common in Ansible module land (though I don't think it's a formalized Guideline). I don't think it's a hard blocker for me in the case where the module states are more native to what a docker swarm user would expect.

**Comment by maxamillion** [comment]
@dagwieers anything else you wanted addressed here? Your review is still pending.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Right I think it's better to use native state (same states as docker swarm).

File: .github/BOTMETA.yml
Comment (dagwieers) [inline .github/BOTMETA.yml]: This is not needed, and even unwanted. If you are listed as the author, you will automatically be considered like this. So we use the BOTMETA.yml as an addendum to what the module already has.

So please remove this.

Comment (dagwieers) [reply .github/BOTMETA.yml]: I have cleaned up the windows and vmware sections, and while they have a massive amount of modules, the actual statements in BOTMETA.yml are minimal. And that's how we want it.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Fine for me. It's not always easy for a reviewer to propose better paradigms, and they do not always exist.

But the legacy of modules doing this wrong is greater than it should be. We ship a lot of bad examples :-)

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: So I noticed this long list of options without defaults or choices. Possibly this is exactly what is intended.
However, in a lot of cases there is an implicit default that could be mentioned (even when it's not enforced when missing).

If you initialize a new swarm I expect there are defaults set.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: No need to use quotes. We tend to not quote when it's not needed, as this helps people understand the YAML rules better.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: No need to quote.

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Like many other statements in the examples.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: .github/BOTMETA.yml
Comment (tbouvet) [reply .github/BOTMETA.yml]: Ok I didn't know. I solve it.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Done

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: What are the default values for te below parameters if you do not specify it on creation ?
- task_history_retention_limit
- keep_old_snapshots
- log_entries_for_slow_followers
- heartbeat_tick
- election_tick
- dispatcher_heartbeat_period
- node_cert_expiry
- ca_force_rotate
- autolock_managers

These possibly require an additional entry in the description to state the defaults on creation. (So you can't add a real default value, because that may modify an existing entry)

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: In the API, all these parameters are optional and by default value is None. They are set only if value is not None. So it's just a docker default value. How do yo suggest to do that ? It's not a default ansible value. By default these parameters are not set and not push to docker API.

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Add the default value to the description. The default values should be documented, if not, create a new swarm and check its value.

How is the module idempotent in this way ? If you would modify a value, how would it know that task made a change or not ?

**Review by dagwieers — APPROVED** [review]
LGTM, I reworked the documentation a bit as I didn't want to cause any more back-and-forth. I hope this is fine. Once this is validated by CI I will merge this.

**Comment by dagwieers** [comment]
@tbouvet Once this is merged (and because you're the author), any subsequent changes/improvements to the module can be auto-merged if you add `shipit` to a PR.

**Comment by tbouvet** [comment]
@dagwieers thanks a lot for all your comments !
I have added a new **update** state to fix idempotent.

**Comment by dagwieers** [comment]
@tbouvet The **update** state is a problem to how people will be using this and it also relate to declarative vs imperative. If you write a playbook, the purpose is to be able to modify it and re-run it.
Which means that the task to create a swarm is the same task that is used for updating swarm settings.

So the **update** state IMO is an unwanted option., the **init** task should be able to create *and* update an existing swarm. (Which we commonly do using state=present)

**Comment by tbouvet** [comment]
@dagwieers Your are right, I have updated states to remove **update**. So now state are: **present/absent/remove/join/inspect**. And idempotent is ok. If no modification are detected in the _present_ state, the return is _change=False_.

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (dagwieers) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: This one I still find confusing. I would expect this would remove/destroy the swarm ?
What is leaving the swarm in this case ? (It may just be terminology and I simply am that ignorant guy)

**Review by dagwieers — APPROVED** [review]
LGTM

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: - **absent** state: remove/destroy the swarm node. So if you put **all** in the _hosts_, it will remove all the swarm.
- **remove** state: if a node leaves a swarm, it is not removed from the swarm configuration,  so this state is to remove the node from the node list.

I think, a lot of cases, **present**/**absent** will be used. The **remove** state will not be used often.

Comment (dagwieers) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Ok, so the description in that example does not match the action, correct ?

Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Problem is that if it's one host, this node leaves the swarm cluster and if the host is all the cluster, the cluster will be destroyed.

**Comment by dagwieers** [comment]
You know you can do sanity tests and integration tests local before you commit, right ?

Something like:
```
./test/runner/ansible-test sanity lib/ansible/modules/cloud/docker/docker_swarm.py
```

**Comment by tbouvet** [comment]
@dagwieers thanks for this help. I will do that now. I have found documentation. Sorry.

**Review by maxamillion — APPROVED** [review]

**Comment by maxamillion** [comment]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by dagwieers** [comment]
merge_rebase

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (samdoran) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: Wrap the contents of `msg` in `to_native` to ensure it's the proper encoding based on the Python version.

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/docker/docker_swarm.py
Comment (maxamillion) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Good catch, thanks @samdoran

Comment (tbouvet) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: @samdoran done !

**Review by samdoran — APPROVED** [review]

**Pull request merged**
